File: Problems/0447_Number_of_Boomerangs/Number_of_Boomerangs.py
# coding: utf-8

import os
import sys
import time
#from collections import defaultdict
import collections

class Solution:
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        ans = 0
        cnt = collections.defaultdict(list)
        for i in range(len(points)-1):
            for j in range(i+1, len(points)):
                dist = (points[i][0] - points[j][0])**2 + (points[i][1] - points[j][1])**2
                cnt[dist].append(set([i, j]))
        for same in cnt.values():
            for i in range(len(same)-1):
                for j in range(i+1, len(same)):
                    if same[i] & same[j]:
                        ans += 2
        return ans

def str_to_int_array(flds):
    if len(flds) <= 0:
        return None
    points = [[0]*2]*len(flds)

    points = [[0 for j in range(2)] for i in range(len(flds))]
    for i in range(len(flds)):
        temp = flds[i].split(",")
        points[i][0] = temp[0]
        points[i][1] = temp[1]
    return points

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    var_str = temp.replace("[[","").replace("]]","").rstrip()
    flds = var_str.split("],[")
    points = str_to_int_array(flds)

    time0 = time.time()

    sl = Solution()
    result = sl.numberOfBoomerangs(points)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()

File: Problems/0447_Number_of_Boomerangs/Number_of_Boomerangs.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public int NumberOfBoomerangs(int[,] points)
    {
        int n = points.GetLength(0);
        int count = 0;

        for (int p0 = 0; p0 < n; p0++)
        {
            // Keep a lookup of the distance from p0 to all other points
            // if you find another point with same distance give that distance
            // a count of 1 (one other point), if you see another point of this
            // distance move count to 2 and so on.  
            Dictionary<int,int> distSqMap = new Dictionary<int,int>();
            for (int p1 = 0; p1 < n; p1++)
            {
                if (p1 == p0) continue;
                
                // avoid square root calculation - do distance check against distance square
                int distSq = (points[p0,0] - points[p1,0])*(points[p0,0] - points[p1,0]) 
                        + (points[p0,1] - points[p1,1])*(points[p0,1] - points[p1,1]);
                
                if (!distSqMap.ContainsKey(distSq))
                {
                    distSqMap[distSq] = 0;
                }
                else
                {
                    distSqMap[distSq]++;
                }
            }
            
            // count number of combinations for groups of equally distanced points
            foreach (int groupCount in distSqMap.Values)
            {
                count += groupCount * (groupCount + 1);
            }
        }    
        
        return count;
    }

    public int[,] str_to_int_array(string[] flds)
    {
        if (flds.Length <= 0)
            return null;

        int[,] points = new int[flds.Length, 2];
        string[] temp;

        for (int i = 0; i < flds.Length; ++i)
        {
            // Console.WriteLine("flds[" + i.ToString() + "] = " + flds[i]);
            temp = flds[i].Split(',');
            points[i,0] = int.Parse(temp[0]);
            points[i,1] = int.Parse(temp[1]);
        }

        return points;
    }

    public string output_points(int[,] points)
    {
        if (points.Length <= 0)
            return "";

        string resultStr = "[[" + points[0,0].ToString() + "," + points[0,1].ToString() + "]";
        // Console.WriteLine("points.Length = " + points.Length);

        for (int i = 1; i < points.Length / 2; ++i)
        {
            resultStr += ",[" + points[i,0].ToString() + "," + points[i,1].ToString() + "]";
            // Console.WriteLine("points[" + i.ToString() + ",0] = " + points[i,0].ToString() + "," + points[i,1].ToString());
        }

        return resultStr;
    }

    public void Main(string args)
    {
        string temp = args.Replace("[[","").Replace("]]","").Trim();
        string[] flds = temp.Split(new string[] {"],["}, StringSplitOptions.None);
        int[,] points = str_to_int_array(flds);

        Console.WriteLine("points = " + output_points(points));
        
        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = NumberOfBoomerangs(points);
        Console.WriteLine("result = " + result.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
        /*
        Console.Write("Hit Any Key");
        Console.Read();
        */
    }
}
